Create utility to generate slugs

BryanHouston opened the pull request:
Addresses https://github.com/openstax/cnx/issues/353, which is part of https://github.com/openstax/cnx/issues/348

mmulich commented on setup.py:
```diff
@@ -6,21 +6,42 @@
 GNU AFFERO GENERAL PUBLIC LICENSE Version 3.0 (AGPL).
 See LICENSE.txt for details.
 """
+import pathlib
```
We should probably use `os.path` here to be backwards compatible with python2, since this will be used in python2 application in the foreseeable future. See also https://github.com/openstax/cnx-easybake/blob/f8edf018fb7499f6f18af0145c326b93a737a782/setup.py#L15-L21

BryanHouston replied:
Fixed.

mmulich commented on tests/test_urlslug.py:
```diff
@@ -0,0 +1,28 @@
+def test_generate_slug():
+    expected_map = {
+        "College Physics / Viscosity and Laminar Flow; Poiseuille’s Law":
+        "/12-4/viscosity-and-laminar-flow-poiseulles-law",
```
Hmm... I can see where this one is going to be interesting, because how would we know this chapter 12 section 4? The title isn't suggesting that at all. 🤔

mmulich reviewed (changes requested):
requestion one small python2 compatibility change

TomWoodward commented on cnxcommon/urlslug.py:
```diff
@@ -0,0 +1,22 @@
+    generates a URL slug.
+
+    GIVEN "Biology 2e / 1.1 The Science of Biology"
+    RETURNS  "/1-1/the-science-of-biology"
```
this should really be 
```
GIVEN "1.1 The Science of Biology"
RETURNS  "1-1-the-science-of-biology"
```

BryanHouston replied:
Fixed.

TomWoodward commented on tests/test_urlslug.py:
```diff
@@ -0,0 +1,72 @@
+
+from cnxcommon.urlslug import generate_slug
+
+def test_generate_slug():
```
do the page titles in the db contain html entities or are those encoded when the archive response is put together? in rex we had to decode html entities _before_ stripping out special characters, but i'm not sure if that will be a concern for you guys.

BryanHouston replied:
Yeah, I'm also working with what _you_ had available. I've updated the code.

`title = re.sub(r"<.*?>", "", title)  # remove html tags`

BryanHouston committed:
Generate URL slugs using unidecode

BryanHouston committed:
Generate slug URLs using slugify

BryanHouston commented:
@pumazi and @tomjw64 - This is ready for review! 😃 Note the differences between each commit. I tested out different python implementation. I started with writing my own by duplication what is currently in `rex-web`, as that was the intention, to make the definitive implementation as closely match the current implementation. Then I made a **small improvement** by using a 3rd-party library called `unidecode` instead of python's built-in `unicodedata`. And finally I eliminated a bunch of lines of my code, all of my own code to match rex-web, actually (😓). So that is what I think we should probably go with unless we want to eliminate a dependency on a 3rd party lib and use own which I've already made and works well, fine, good enough at least on all the tests I wrote/ran.

BryanHouston committed:
Generate URL slugs also for chapter titles

BryanHouston committed:
Add integration test based on acceptance criteria for generating slug URLs

BryanHouston commented:
@brittweinstein Note that I'm unable to add a milestone to this PR; I don't know if it matters or how much.

mmulich commented on README.md:
```diff
@@ -2,6 +2,12 @@
+
+## urlslug
+
+Generates a URL slug for REX. # TODO
```
☝️ `TODO`?

BryanHouston replied:
😬

BryanHouston replied:
Fixed.

mmulich commented on tests/test_urlslug.py:
```diff
@@ -0,0 +1,152 @@
+            ('biology-2e', '1-1-the-science-of-biology'),
+            ('biology-2e', 'preface')
+        ]
+
```
We go from returning one type `str` in some situations to a `tuple` in other situations. How about we only return a `str`?

mmulich replied:
The reason for this would be something along the lines for functional programming expectations. While we can do this kind of dynamic return in Python, that doesn't mean we should. Python isn't strictly typed but that's not to say we should ignore those practices in usage.

BryanHouston replied:
Yeah, I agree. However, I would've thought you'd want the opposite. First option: you call the function twice – once (always) for generating the chapter title and *then* the section title. Second option: you call the function once. If you provide only one argument, then you get a tuple with one item, but if you call it with 2-3 arguments, you get a tuple with 2 items every time. So I'll go with the second option: to return a tuple always... with 1 OR 2 items.

BryanHouston replied:
Done.

mmulich commented on cnxcommon/urlslug.py:
```diff
@@ -0,0 +1,58 @@
+        # find chapter number in chapter title
+        chapter_number = re.split(r"^([0-9\.]*)?(.*)$", chapter_title)[1]
+        # update section title with the chapter number, only if found.
+        section_title = f"{chapter_number}-{section_title}" if chapter_number else f"{section_title}"
```
Very good use of format string, but unfortunately it's a Python3 only feature. To make this backwards compatible you're going to need to use `<str>.format(...)`.

BryanHouston replied:
OH. 😣

BryanHouston replied:
Fixed.

BryanHouston committed:
Adress PR suggestions

BryanHouston commented:
@pumazi Addressed all of your PR comments/suggestions. Pushed in my most recent commit.

mmulich commented on cnxcommon/requirements/main.txt:
```diff
@@ -0,0 +1,2 @@
+# The project's main runtime dependencies
+slugify==3.0.2
```
should be `python-slugify`. You can get the name and version from `pip freeze`

BryanHouston replied:
Fixed.

mmulich reviewed (dismissed):
Please put the requirements directory at the root of the project. `setup.py` as written is broken due to the path to these files not resolving.

mmulich commented:
```pytb
[INFO] dbmigrator (20190424172324_path_slug_field_for_baked_books.py) - processing... 17661524 - ['Principles of Management - DRAFT']
Traceback (most recent call last):
  File "/usr/local/bin/dbmigrator", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python2.7/dist-packages/dbmigrator/cli.py", line 116, in main
    return args['cmmd'](**args)
  File "/usr/local/lib/python2.7/dist-packages/dbmigrator/utils.py", line 155, in wrapper
    return func(cursor, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/dbmigrator/commands/migrate.py", line 32, in cli_command
    run_deferred)
  File "/usr/local/lib/python2.7/dist-packages/dbmigrator/utils.py", line 237, in compare_schema
    callback(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/dbmigrator/utils.py", line 270, in run_migration
    migration.up(cursor)
  File "../../src/cnxdb/migrations/20190424172324_path_slug_field_for_baked_books.py", line 83, in up
    slug = generate_slug(title)
  File "/tmp/cnx-common/cnxcommon/urlslug.py", line 38, in generate_slug
    book_title = slugify(remove_html_tags(book_title))
  File "/tmp/cnx-common/cnxcommon/urlslug.py", line 59, in remove_html_tags
    return re.sub(r"<.*?>", "", title)
  File "/usr/lib/python2.7/re.py", line 155, in sub
    return _compile(pattern, flags).sub(repl, string, count)
TypeError: expected string or buffer
```
The calling code for this would roughly be: `generate_slug(['Principles of Management - DRAFT'])`

BryanHouston commented:
@pumazi You have to call it this way: `generate_slug(*['Principles of Management - DRAFT'])`

BryanHouston commented:
@pumazi Addressed all your suggestions moving the requirements directory to the root of the project.

BryanHouston commented:
Per our last conversation in a call...
    
     ++ Requirement: The function should always return a single string.

## When given 2-3 parameters
- Try to build the complete section title slug from the last parameter.
- If the last parameter does not begin with a number, then (chapter number is missing, so) look at second from last parameter to build the complete section title slug.
    - Using the second from last parameter, try to determine the chapter number.
    - Build the complete section title slug: chapter number + section title like `1-introduction`
    - Return complete **section title** slug (`string`)

## When given 1 parameter:
- Build book title slug
- Return **book title** slug (`string`)

BryanHouston committed:
Slug generator - return string for book OR section title

Previously, the slug generator returned a tuple with 1-2 items... the book title slug AND the section title slug.

After these changes, it always returns a string and uses the second item only if the complete section title slug can't be built because it's missing the chapter title in the 3rd parameter.

BryanHouston commented:
@pumazi Pushed a commit, ready for review.

BryanHouston committed:
Create slug generator util for REX
